content/posts/entendendo-rxjs-observable-com-angular.md

```diff
@@ -0,0 +1,202 @@
+date: 2017-08-08
+excerpt: Uma forma simples, com exemplos reais de como utilizar RxJS e aplicar conceitos de programação reativa com Angular
+categories:
+  - JavaScript
+  - Tecnologia e Tendências
+  - AngularJS
+---
+
+Após passar por um projeto com *Angular 2*(ou somente *Angular*, para os mais íntimos) posso dizer que: É um *framework* com muitas vantagens, e uma das coisas mais fascinantes que conheci foi uma biblioteca maravilhosa que é o *[RxJS](https://github.com/Reactive-Extensions/RxJS)*(estude este cara e verá as maravilhas de um mundo moderno) e conheci este cara que mudou a minha vida, o famoso ***Observable***.
+
+### O que é um Observable?
+
+Por definição é uma coleção que funciona de forma unidirecional, ou seja, ele emite notificações sempre que ocorre uma mudança em um de seus itens e a partir disto podemos executar uma ação. Digamos que ele resolve o mesmo problema que a versão anterior do *Angular* havia resolvido com o *$watch*, porém sem usar força bruta. Enquanto no *$watch* verificamos todo nosso escopo por alterações após cada *$digest cycle*(o que tem um grande custo na performance), com *Observable* esta verificação não acontece, pois para cada evento é emitida uma notificação para nosso *Observable* e então tratamos os dados.
+
+### O que faço com Observable? E onde entra Angular nessa história?
+
+Vamos imaginar o consumo de uma *web service*, algo bem comum em aplicação *Single Page*, antes do Observable faríamos dessa maneira
+
+    getUsers(): Promise<User[]>{
+      return fetch(myApiUrl)
+      .then(res=>res.json())
+      .catch(err=>{
+         throw new Error(err.message);  
+       });
+    }
+
+O código acima é bem simples, estamos buscando um recurso(Uma lista de usuários, neste caso) e após a resposta transformamos tudo em *JSON.*
+
+Entretanto, o novo Angular foi construído pensando em aplicações reativas, abandonando o uso de *Promises* e adotando por padrão o ***Observable***.
+
+Usando Observable a mesma função ficaria da seguinte maneira:
+
+    @Injectable()
+    class UserService {
+     ...
+     getUsers(): Observable<User[]> {
+        return this.http.get(myApiUrl)
+                        .map(res=>res.json())
+                        .catch(err=>{
+                           throw new Error(err.message);  
+                        });
+     } 
+     ...
+    }
+
+### Mas quais são as vantagens de usar Observable e não Promises?
+
+A grande vantagem está nos “poderes” que o *Observable* nos dá com seus **operadores**, por exemplo: Podemos “cancelar” *requests* para poupar processamento, ou até mesmo tentar fazer uma nova requisição caso algum problema como perda de conexão aconteça.
+
+O usuário não precisa ver aquela tela de erro, afinal, ele não tem culpa.
+
+### Operador Subscribe
+
+Um dos operadores mais importantes do *Observable* é o *subscribe*, que é o equivalente ao *then* de uma *promise* ou ao *$watch*.
+
+    import { UserService } from 'user.service';
+
+    @Component({
```

Comment:
Todas as partes do código são relevantes? o que não for, pode tirar fora.

Reply:
Acho que são importantes para dar o contexto de que usamos dentro de um componente e de onde estamos chamando o método, que no caso é da service. Acha que vale a pena dar uma enxugada e deixar só o trecho relativo ao operador?

Reply:
Se fica mais fácil de entender com todo o código, pode deixar.